server/android_notification_server.go

```diff
@@ -166,16 +166,9 @@ func (me *AndroidNotificationServer) SendNotification(msg *PushNotification) Pus
 		},
 	}
 
-	ctx, cancel := context.WithTimeout(context.Background(), me.sendTimeout)
-	defer cancel()
-
 	me.logger.Infof("Sending android push notification for device=%v type=%v ackId=%v", me.AndroidPushSettings.Type, msg.Type, msg.AckID)
 
-	start := time.Now()
-	_, err := me.client.Send(ctx, fcmMsg)
-	if me.metrics != nil {
-		me.metrics.observerNotificationResponse(PushNotifyAndroid, time.Since(start).Seconds())
-	}
+	err := me.SendNotificationWithRetry(fcmMsg, 0)
```

Comment:
For Android, retries are already built-in to the HTTP client: https://github.com/firebase/firebase-admin-go/blob/6a28190d0d30fe0134648566c7e89f8635a51672/messaging/messaging.go#L892.

There is no need to do anything. In future, please feel free to ask me for review or take a deeper look at the documentation. :)

Reply:
I missed that. Thank you for pointing out!

That being said, the reason we are retrying is, among others, because FCM may return a "internal server error", which looking around the web, seems to be a fluke in their system, and that we just have to retry.

Looking at the internals, I see that only `serviceUnavailable` status is retried, so other problems (like the context timing out, or this internal server error) won't get retried. Am I right?

If that is so, how would you approach this?

Reply:
I would trust on the implementation to be correct on which status codes to retry on. Also, any network errors like timeouts are also retried.

So the only missing thing is to retry on additional status codes. I'd want to see some metrics on the distribution of errors we are getting from FCM and then take a call. 

For context, overriding the retry policy from the user side is difficult: https://github.com/firebase/firebase-admin-go/issues/579.

Reply:
Right now I don't see FCM errors on the metrics, but @marianunez spotted a few some time ago.

For iOS we recently have found a couple regarding context timeout and others, so that is why we are adding the retry for both.

Reply:
For example, in the last 7 days, there has been 54 instances of FCM failures: 

<img width="1703" alt="Screenshot 2024-08-22 at 8 30 26 AM" src="https://github.com/user-attachments/assets/bffe7095-83fa-4cdb-a9f1-f8929429bf03">

Reply:
Now I see why I didn't see many errors on android lately... I was looking at the metrics... of apple :facepalm: 

The number of errors are not huge in terms of percents, but still they would mean non-sent notifications, so we should avoid them.

Reply:
> Now I see why I didn't see many errors on android lately... I was looking at the metrics... of apple 

It would be alarming to see FCM issues coming from the APNS 🤣

Reply:
Maria - from your screenshot, it looks like a lot of them are "Requested entity was not found" which we should not be retrying on. Do you have access to the Grafana dashboard for this? I'd like to see the distribution of errors in the last 7 days.

And on the side, I think we should recover the actual status code from the `internal.FirebaseError` and start logging/graphing that. Only then we will get a better picture of what other status codes are we missing out on. For now, everything is bucketed under "internal error".

I'd suggest we do this, gather the data, and then come up with a retry strategy for Android. For iOS, we can go ahead with the PR changes. Because the client does not allow us to easily override the retry policy, so it will take some effort.

Reply:
@agnivade Not everything is bucketed as internal error. In the push proxy metrics we make the distinction here: https://github.com/mattermost/mattermost-push-proxy/blob/9a50bf22d81777d8ed47ddecf189bbf6a386163e/server/android_notification_server.go#L192
Each error is bucketed differently, and many are "expected errors". But internal error seems to be a hiccup on FCM side.

Reply:
Sorry, I meant that all different 5xx errors are bucketed as internal error. I want to understand what is the distribution of HTTP status codes we get from FCM and make a data based decision.

Reply:
Not sure how we can really get that. The library exposes just an error, and doesn't expose the http status code. We could use some castings to get it out, but not sure if that is what we want.

We also log the `Failed to send FCM push` logs... But those do not have the status code either. They are already "filtered" by the library.

Regarding accessing to those logs... you would need access to the production logs, that probably you don't have. I am able to access the Grafana Dashboard (on `Mattermost Release` workspace, not `Mattermost Developers`), but I am not able to inspect the logs. That being said, @marianunez can help you get the logs you may need.

Reply:
> Do you have access to the Grafana dashboard for this? I'd like to see the distribution of errors in the last 7 days.

@agnivade These are the Loki logs for the prod push proxy. In 7 days, of 54 entries of `Failed to send FCM push`:

- 34 were `Internal error encountered `
- 18 were `Requested entity was not found`

Reply:
FWIW, [the documentation ](https://firebase.google.com/docs/cloud-messaging/http-server-ref#error-codes) clearly indicates that the sender must retry for 5xx errors

<img width="1164" alt="Screenshot 2024-08-23 at 12 14 39 PM" src="https://github.com/user-attachments/assets/0226606f-f7c0-45fd-b7bd-00b26e677ad2">

@larkox  Also, we need to implement the exponential back-off as I mentioned above.

Reply:
@marianunez The library doesn't expose the error codes. It has a few util functions to distinguish the errors, and that is what we are doing in the PR.

Reply:
> Not sure how we can really get that. The library exposes just an error, and doesn't expose the http status code. We could use some castings to get it out, but not sure if that is what we want.

@larkox - casting is exactly what we want :)

@marianunez - thanks for the link. It's clear that we should retry on 5xx. However, from the logs, it's not clear if the ones with internal error were logged after a retry with 503, or they were logged from some other non-503 error which wasn't retried. That's what I am trying to find out here.

In any case, the library is poorly written and makes it such that the user cannot choose to override the retry policy inspite of the fact that the client clearly takes a retry policy struct as a parameter. There isn't even a way to bypass retrying so that we can copy over just the retry code.

So in worst case, we have to fork the repo and move out some packages out of internal, so that they can be imported and overridden from the client.

But before that, I wanted some metrics on the distribution of the status codes. If we see that out of all 34 internal errors, all were actually 503, then I think we could leave this alone. But if we see other non-503 status codes, then we have a decision in our hand to make.

Reply:
@agnivade Just to understand your ask.

You want a first change to surface the status code of the response, let it simmer for a while so we get new data, verify what kind of 5xx error it is, and then decide whether this PR makes sense or not. Am I right?

Reply:
@larkox - Yep, exactly.

@marianunez - To clarify: we, ourselves, don't want to retry for 503, because the client already does by default. My objective here is to figure out if the internal failures are all 503 or not. If it is, then we know that those were already retried and still failed. If it's something other than 503, then yes we definitely need to do something ourselves.